### PR TITLE
[IA-4448] azure autopause integration test

### PR DESCRIPTION
**Pull request by lucymcnatt**
__Jira ticket__: https://broadworkbench.atlassian.net/browse/[ticket_number]

### What

-

### Why

-

### What to test

- 

### Who tested and where

- [ ] This change is covered by automated tests 
  - _NB: Rerun automation tests on this PR by commenting `jenkins retest` or `jenkins multi-test`._
- [ ] I validated this change 
- [ ] Primary reviewer validated this change 
- [ ] I validated this change in the dev environment

### CODE REVIEW

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        createReq = new CreateAzureRuntimeRequest()
+          .labels(labelMap)
+          .machineSize("Standard_DS1_v2")
+          .disk(
+            new AzureDiskConfig()
+              .name(generateAzureDiskName())
+              .size(50)
+              .labels(labelMap)
+          )
+          .autoPauseThreshold(15)
```
Comment (jdcanas) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala]: unit comment could be helpful

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        // new IO for new stream
+        callGetRuntimeStopping = IO(runtimeClient.getAzureRuntime(workspaceId, runtimeName.asString))
+
+        _ <- loggerIO.info(
+          s"AzureAutoPauseSpec: runtime $workspaceId/${runtimeName.asString} waiting to autopause"
+        )
+
+        monitorAutoPauseResult <- streamUntilDoneOrTimeout(
+          callGetRuntimeStopping,
+          240,
```
Comment (jdcanas) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala]: Can this be lower? This doesn't depend on WSM timeouts like the other polling steps, so we should be able to use something derived from the specified autopause threshold + leo's monitoring interval?

Comment (lucymcnatt) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala]: yeah I can shorten this, I had it that high while testing

File: automation/README.md [hunk 3]
```diff
@@ -1,5 +1,25 @@
...
+```
+
+**Arguments:** (arguments are positional)
+
+* branch of firecloud-automated-testing
+	* Configs branch; defaults to `master`
+* Vault auth token
+	* Defaults to reading it from the .vault-token via `$(cat ~/.vault-token)`.
+* env
+	* Environment of your FiaB; defaults to `dev`
```
Comment (LizBaldo) [inline automation/README.md]: FiaB or BEE? I guess the FiaB for automation test is really a BEE under the hood

File: automation/README.md [hunk 4]
```diff
@@ -1,5 +1,25 @@
 Quickstart: running swintegration tests locally on Mac/Docker 
 
+## Set Up
```
Comment (LizBaldo) [inline automation/README.md:3]: Did this set up work for you?

File: automation/README.md [hunk 5]
```diff
@@ -53,7 +58,7 @@ Note: If the test you're trying to run is annotated with `@DoNotDiscover`, do th
 
 ### Developing azure automation tests locally
 
-When running azure automation tests locally against a fresh [bee](https://broadworkbench.atlassian.net/wiki/spaces/IA/pages/104399223/Callisto+Developer+Handbook#Running-in-a-BEE), you will have to perform a few extra steps:
+When running azure automation tests locally against a fresh [bee](https://broadworkbench.atlassian.net/wiki/spaces/IA/pages/2839576631/How+to+BEE), you will have to perform a few extra steps:
```
Comment (LizBaldo) [inline automation/README.md:61]: Thanks!

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala (hunk 1 above)
Comment (LizBaldo) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala]: Might be good to double check with Lily that 15 minutes is still the desired duration for testing

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+          )
+          .autoPauseThreshold(15)
+
+        _ <- IO(runtimeClient.createAzureRuntime(workspaceId, runtimeName.asString, false, createReq))
+        _ <- loggerIO.info(s"AzureAutoPauseSpec: Create runtime request submitted. Starting to poll GET")
+
+        // ------------------ Waiting for Create ------------------ //
+
+        // Verify the initial getRuntime call
+        _ <- IO.sleep(5 seconds)
```
Comment (LizBaldo) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala]: Arg, i keep forgetting why we need a `sleep` statement here, could you add a little comment?

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        )(implicitly, GeneratedLeonardoClient.runtimeInStateOrError(ClusterStatus.RUNNING))
+
+        _ <- loggerIO.info(
+          s"AzureAutoPauseSpec: runtime ${workspaceId}/${runtimeName.asString} create monitor result: $monitorCreateResult"
+        )
+        _ = monitorCreateResult.getStatus() shouldBe ClusterStatus.RUNNING
+
+        // ------------------ Waiting for Autopause ------------------ //
+
+        // new IO for new stream
```
Comment (LizBaldo) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala:88]: Is this to avoid the weird exit observed by Justin on like 5% of the tests?

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        _ <- loggerIO.info(
+          s"AzureAutoPauseSpec: runtime $workspaceId/${runtimeName.asString} waiting for runtime to stop"
+        )
+
+        // new IO for new stream
+        callGetRuntimeStopped = IO(runtimeClient.getAzureRuntime(workspaceId, runtimeName.asString))
+
+        monitorStoppingResult <- streamUntilDoneOrTimeout(
+          callGetRuntimeStopped,
+          60,
```
Comment (LizBaldo) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala:119]: We could probably have a shorter timeout here as well?

File: http/src/main/resources/swagger/api-docs.yaml [hunk 9]
```diff
@@ -3447,6 +3447,9 @@ components:
           description: an optional set of key value pairs for environment variables to be injected into the VM
         disk:
           $ref: "#/components/schemas/AzureDiskConfig"
+        autoPauseThreshold:
```
Comment (LizBaldo) [inline http/src/main/resources/swagger/api-docs.yaml]: Would it make sense to separate this into its own PR to publish the real version you can use for good in this PR? Just thinking, I am not very opinionated :)

**Review by LizBaldo — APPROVED** [review]
LGTM!

I left a few nit comments around documenting a few steps / potentially breaking one change out into its own PR.

If it runs for you locally I would say go for it and test in the GHA 🤞 you might be able to modify the yml file to have this test run as a PR commit?

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala (hunk 7 above)
Comment (lucymcnatt) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala:88]: yup

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala (hunk 6 above)
Comment (lucymcnatt) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala]: I'll remove

File: http/src/main/resources/swagger/api-docs.yaml (hunk 9 above)
Comment (lucymcnatt) [reply http/src/main/resources/swagger/api-docs.yaml]: hmm I considered that, but I don't really see the harm in keeping it here + opening a new leo PR and waiting on that merge might slow things down

File: core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/azureRoutesModels.scala [hunk 10]
```diff
@@ -6,7 +6,7 @@ final case class CreateAzureRuntimeRequest(labels: LabelMap,
                                            machineSize: VirtualMachineSizeTypes,
                                            customEnvironmentVariables: Map[String, String],
                                            azureDiskConfig: CreateAzureDiskRequest,
-                                           autoPauseThreshold: Option[Int]
+                                           autopauseThreshold: Option[Int]
```
Comment (lucymcnatt) [inline core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/azureRoutesModels.scala:9]: standardized capitalization for autopause

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        // ------------------ Waiting for Autopause ------------------ //
+
+        // new IO for new stream
+        callGetRuntimeStopping = IO(runtimeClient.getAzureRuntime(workspaceId, runtimeName.asString))
+
+        _ <- loggerIO.info(
+          s"AzureAutoPauseSpec: runtime $workspaceId/${runtimeName.asString} waiting to autopause"
+        )
+
+        _ <- IO.sleep(20 minutes) // sleep for 20 minutes before checking if the runtime paused
```
Comment (lucymcnatt) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala]: I found that it wasn't autopausing when I polled for some reason...but I also changed a couple other things between so I can try adding that back and testing again during my meetings today

File: .github/workflows/azure_automation_test.yml [hunk 12]
```diff
@@ -131,10 +131,18 @@ jobs:
           token: ${{ env.TOKEN }}
           inputs: '{ "bee-name": "${{ env.BEE_NAME }}" }'
 
+
+#  report-workflow:
```
Comment (LizBaldo) [inline .github/workflows/azure_automation_test.yml]: I don't remember, but do we need this? If not, should it be removed?

File: automation/README.md (hunk 4 above)
Comment (LizBaldo) [reply automation/README.md:3]: I am really curious actually because it did not work for me 🤔

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSuite.scala [hunk 13]
```diff
@@ -431,8 +431,10 @@ final class LeonardoTerraDockerSuite
 
 final class LeonardoAzureSuite
     extends Suites(
-      new AzureRuntimeSpec,
-      new AzureDiskSpec
+      // will remove after testing
```
Comment (LizBaldo) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSuite.scala]: Do you need to remove now?

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala (hunk 11 above)
Comment (LizBaldo) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureAutopauseSpec.scala]: Yeah, I would add polling, because the listener is supposed to ignore that

File: project/Dependencies.scala [hunk 14]
```diff
@@ -224,7 +224,7 @@ object Dependencies {
   )
 
   val workbenchServiceTest: ModuleID = "org.broadinstitute.dsde.workbench" %% "workbench-service-test" % serviceTestV % "test" classifier "tests" excludeAll (excludeGuava, excludeStatsD)
-  val leonardoClient: ModuleID =  "org.broadinstitute.dsde.workbench" %% "leonardo-client" % "1.3.6-563edbd-SNAP"//"1.3.6-9d5d754"
+  val leonardoClient: ModuleID =  "org.broadinstitute.dsde.workbench" %% "leonardo-client" % "1.3.6-8366c9f-SNAP"
```
Comment (LizBaldo) [inline project/Dependencies.scala:225]: I updated this am and you should not need the SNAP version anymore

File: .github/workflows/azure_automation_test.yml (hunk 12 above)
Comment (lucymcnatt) [reply .github/workflows/azure_automation_test.yml]: ohh yeah this was when I was testing different options for the slack notification, removed

File: project/Dependencies.scala (hunk 14 above)
Comment (LizBaldo) [reply project/Dependencies.scala:225]: Oh yeah sorry I keep forgetting you need that to get the autoPause parameter in the client!

**Review by LizBaldo — APPROVED** [review]
